### PR TITLE
replace DavClient.js with our own implementation, optimize parsers

### PR DESCRIPTION
**Pull request by georgehrke**
This pull-request:
- removes davclient.js as dependency (we used only a fraction of it)
- properly rejects Promises if http status code is not 200-299
  (@skjnldsv fixes your issue not to send find up on unsuccessful create / update)
- moves parsing out of models into dedicated class that's called by Request
- we now use XPath to parse the data, which is faster than our previous approach.
- adds unit test for everything except index.js

---

- I removed the parser for {DAV:}acl, because it's a pain to parse and we don't actually need it. all we need to know is inside {DAV:}current-user-privilege-set

### CODE REVIEW

File: src/debug.js [hunk 1]
```diff
@@ -29,7 +29,8 @@
 export function debugFactory(context) {
 	return (...args) => {
 		if (debugFactory.enabled) {
-			console.debug(context, args);
+			// eslint-disable-next-line no-console
```
Comment (skjnldsv) [inline src/debug.js]: You should use debug since it's ...a debug :grin:

**Comment by georgehrke** [comment]
@skjnldsv Please test :)

**Comment by georgehrke** [comment]
(Really gotta fix the coverage report to put us a nice 95% coverage badge in our Readme :P)

**Comment by georgehrke** [comment]
@skjnldsv Did you have any chance to review this yet? I need this for the sharing PR. 🙂

**Comment by skjnldsv** [comment]
@georgehrke I'm on vacation :see_no_evil:

**Comment by georgehrke** [comment]
Oh, sorry for bothering. Enjoy your vacation and don't you dare to test it while being on vacation :P

File: package.json [hunk 2]
```diff
@@ -63,9 +63,6 @@
 		"webpack-merge": "^4.1.4"
 	},
 	"dependencies": {
-		"@babel/polyfill": "^7.0.0",
-		"compare-urls": "^2.0.0",
-		"davclient.js": "git+https://github.com/owncloud/davclient.js.git",
-		"normalize-url": "^3.3.0"
+		"@babel/polyfill": "^7.0.0"
```
Comment (skjnldsv) [inline package.json]: Feels great! :D

File: src/index.js [hunk 3]
```diff
@@ -37,23 +38,81 @@ export default class DavClient {
 
 	/**
 	 * @param {Object} options
+	 * @property {String} rootUrl
 	 * @param {Function} xhrProvider
 	 * @param {Object} factories
 	 */
 	constructor(options, xhrProvider = null, factories = {}) {
```
Comment (skjnldsv) [inline src/index.js]: No `rootUrl` in the constructor?

File: src/models/davCollection.js [hunk 4]
```diff
@@ -42,19 +41,21 @@ export class DavCollection extends DAVEventListener {
 	constructor(parent, request, url, props) {
 		super();
 
+		// This is a collection, so always make sure to end with a /
```
Comment (skjnldsv) [inline src/models/davCollection.js]: :+1:

File: src/models/davCollection.js [hunk 5]
```diff
@@ -121,21 +125,28 @@ export class DavCollection extends DAVEventListener {
...
 		const [skeleton, dPropChildren] = XMLUtility.getRootSkeleton(
 			[NS.DAV, 'mkcol'],
 			[NS.DAV, 'set'],
 			[NS.DAV, 'prop']
 		);
 
-		props.forEach((prop) => {
-			dPropChildren.push(prop);
-		});
+		dPropChildren.push(...props);
```
Comment (skjnldsv) [inline src/models/davCollection.js]: Nice!

**Comment by skjnldsv** [comment]
@georgehrke I found a bug!
https://github.com/nextcloud/cdav-library/blob/master/src/models/davObject.js#L117

This returns undefined now.
https://github.com/nextcloud/contacts/blob/134527baa933af63400c285b362d9ed889abb001/src/store/contacts.js#L281

**Comment by georgehrke** [comment]
Ah, sorry. I should have documented that better! 🙈 
The Promise will only resolve if the response code is in the 200-299 section.
400+ response codes need to be caught via catch.

This should do the trick:
```
			return contact.dav.update()
				.then((response) => {
					// all clear, let's update the store
					context.commit('updateContact', contact)
				})
				.catch((error) => {
					// wrong etag, we most likely have a conflict
					if (error.status === 412) {
						// saving the new etag so that the user can manually
						// trigger a fetchCompleteData without any further errors
						contact.conflict =error.xhr.getResponseHeader('etag')
					} else {
						...
					})
```

**Pull request merged**